File: dynabuffers-python/dynabuffers/DynabuffersVisitor.py
from dynabuffers.antlr.DynabuffersParser import DynabuffersParser
from dynabuffers.antlr.DynabuffersVisitor import DynabuffersVisitor as DynabuffersBaseVisitor
from dynabuffers.ast.ClassType import ClassType, ClassTypeOptions
from dynabuffers.ast.EnumType import EnumType, EnumTypeOptions
from dynabuffers.ast.FieldType import FieldType, FieldTypeOptions
from dynabuffers.ast.NamespaceType import NamespaceType, NamespaceTypeOptions
from dynabuffers.ast.UnionType import UnionType, UnionTypeOptions
from dynabuffers.ast.datatype.ArrayType import ArrayType, ArrayTypeOptions
from dynabuffers.ast.datatype.BooleanType import BooleanType
from dynabuffers.ast.datatype.ByteType import ByteType
from dynabuffers.ast.datatype.FloatType import FloatType
from dynabuffers.ast.datatype.IntType import IntType
from dynabuffers.ast.datatype.LongType import LongType
from dynabuffers.ast.datatype.MapType import MapType, MapTypeOptions
from dynabuffers.ast.datatype.OptionType import OptionType, OptionTypeOptions
from dynabuffers.ast.datatype.RefType import RefType, RefTypeOptions
from dynabuffers.ast.datatype.ShortType import ShortType
from dynabuffers.ast.datatype.StringType import StringType, StringTypeOptions
from dynabuffers.ast.structural.Annotation import AnnotationOptions, Annotation
from dynabuffers.ast.structural.ClassOptions import ClassOptions, ClassOptionsOptions
from dynabuffers.ast.structural.FieldOptions import FieldOptions, FieldOptionsOptions
from dynabuffers.ast.structural.UnionOptions import UnionOptions, UnionOptionsOptions
from dynabuffers.ast.structural.Value import Value, ValueOptions


class DynabuffersVisitor(DynabuffersBaseVisitor):

    def visitEnumType(self, ctx: DynabuffersParser.EnumTypeContext):
        name = str(ctx.getChild(1))
        values = []
        for index in range(3, ctx.getChildCount() - 1):
            values.append(str(ctx.getChild(index)))

        return EnumType(EnumTypeOptions(name, values))

    def visitClassType(self, ctx: DynabuffersParser.ClassTypeContext):
        values = super().visitClassType(ctx)
        options = next(filter(lambda x: isinstance(x, ClassOptions), values), ClassOptions(ClassOptionsOptions(False, False, False)))

        fields = list(filter(lambda x: isinstance(x, FieldType), values))
        return ClassType(ClassTypeOptions(ctx.getChild(1).getText(), fields, options))

    def visitUnionType(self, ctx: DynabuffersParser.UnionTypeContext):
        values = []
        for i in range(3, ctx.getChildCount() - 1):
            # TODO: remove workaround
            if str(ctx.getChild(i)) != "{":
                values.append(str(ctx.getChild(i)))

        options = next(filter(lambda x: isinstance(x, UnionOptions), values), UnionOptions(UnionOptionsOptions(False, False, False)))

        return UnionType(UnionTypeOptions(str(ctx.getChild(1)), values, options))

    def visitFieldType(self, ctx: DynabuffersParser.FieldTypeContext):
        values = super().visitFieldType(ctx)
        annotations = list(filter(lambda x: isinstance(x, Annotation), values))
        name = ctx.getChild(len(annotations)).getText()
        datatype = values[len(annotations)]
        options = next(filter(lambda x: isinstance(x, FieldOptions), values), FieldOptions(FieldOptionsOptions(False)))
        defaultVal = next(map(lambda x: x.options.value, filter(lambda x: isinstance(x, Value), values)), None)

        return FieldType(FieldTypeOptions(name, datatype, options, annotations, defaultVal))

    def visitDataType(self, ctx: DynabuffersParser.DataTypeContext):
        if ctx.getText() == "string":
            return StringType(StringTypeOptions("utf-8"))
        if ctx.getText() == "float":
            return FloatType()
        if ctx.getText() == "int":
            return IntType()
        if ctx.getText() == "boolean":
            return BooleanType()
        if ctx.getText() == "long":
            return LongType()
        if ctx.getText() == "byte":
            return ByteType()
        if ctx.getText() == "short":
            return ShortType()
        if ctx.getText() == "map":
            return MapType(MapTypeOptions("utf-8"))
        else:
            return RefType(RefTypeOptions(ctx.getText()))

    def visitArrayType(self, ctx: DynabuffersParser.ArrayTypeContext):
        datatype = super().visitArrayType(ctx)[0]
        return ArrayType(ArrayTypeOptions(datatype))

    def visitOptionType(self, ctx: DynabuffersParser.OptionTypeContext):
        datatype = super().visitOptionType(ctx)[0]
        return OptionType(OptionTypeOptions(datatype))

    def visitClassOptions(self, ctx: DynabuffersParser.ClassOptionsContext):
        implicit = "implicit" in ctx.getText()
        return ClassOptions(ClassOptionsOptions("primary" in ctx.getText(), "deprecated" in ctx.getText(), implicit))

    def visitUnionOptions(self, ctx: DynabuffersParser.UnionOptionsContext):
        implicit = "implicit" in ctx.getText()
        return UnionOptions(UnionOptionsOptions("primary" in ctx.getText(), "deprecated" in ctx.getText(), implicit))

    def visitFieldOptions(self, ctx: DynabuffersParser.FieldOptionsContext):
        return FieldOptions(FieldOptionsOptions("deprecated" in ctx.getText()))

    def visitAnnotation(self, ctx: DynabuffersParser.AnnotationContext):
        values = super().visitAnnotation(ctx)
        return Annotation(AnnotationOptions(ctx.getChild(1).getText(), values))

    def visitNamespaceType(self, ctx: DynabuffersParser.NamespaceTypeContext):
        _name = ctx.getChild(1).getText()
        _list = list(filter(lambda x: not isinstance(x, NamespaceType), super().visitNamespaceType(ctx)))
        _nestedNamespaces = list(filter(lambda x: isinstance(x, NamespaceType), super().visitNamespaceType(ctx)))
        return NamespaceType(NamespaceTypeOptions(_name, _list), _nestedNamespaces)

    def visitValue(self, ctx: DynabuffersParser.ValueContext):
        if ctx.getText() == "[]":
            return Value(ValueOptions([]))
        if ctx.getText() == "[:]":
            return Value(ValueOptions({}))
        return Value(ValueOptions(ctx.getText()))

    def aggregateResult(self, aggregate, nextResult):
        array = []
        if aggregate is not None:
            array += aggregate
        if nextResult is not None:
            array += [nextResult]
        return array

File: dynabuffers-java/src/main/kotlin/dynabuffers/DynabuffersVisitor.kt
/*
 * Copyright (c) 2016-2018, Leftshift One
 * __________________
 * [2018] Leftshift One
 * All Rights Reserved.
 * NOTICE:  All information contained herein is, and remains
 * the property of Leftshift One and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Leftshift One
 * and its suppliers and may be covered by Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Leftshift One.
 */

package dynabuffers

import dynabuffers.antlr.DynabuffersBaseVisitor
import dynabuffers.antlr.DynabuffersParser
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.ast.*
import dynabuffers.ast.EnumType.EnumTypeOptions
import dynabuffers.ast.datatype.*
import dynabuffers.ast.structural.*
import dynabuffers.ast.structural.Annotation
import org.antlr.v4.runtime.ParserRuleContext
import java.nio.charset.Charset
import java.util.*
import kotlin.collections.ArrayList

class DynabuffersVisitor(private val charset: Charset) : DynabuffersBaseVisitor<List<IType>>() {

    override fun visitEnumType(ctx: DynabuffersParser.EnumTypeContext): List<IType> {
        return listOf(EnumType(EnumTypeOptions(ctx.textAt(1), ctx.textAt(3, 2), charset)))
    }

    override fun visitUnionType(ctx: DynabuffersParser.UnionTypeContext): List<IType> {
        val list = super.visitUnionType(ctx)
        val unionOptions = (list.firstOrNull { it is UnionOptions }
                ?: UnionOptions(UnionOptions.UnionOptionsOptions(false, false, false))) as UnionOptions

        // TODO: remove workaround
        return listOf(UnionType(UnionType.UnionTypeOptions(ctx.textAt(1), ctx.textAt(3, 2)
                .filter { it != "{" && it != "}" }, unionOptions)))
    }

    override fun visitFieldType(ctx: DynabuffersParser.FieldTypeContext): List<IType> {
        val list = super.visitFieldType(ctx)

        val annotations = list.filter { it is Annotation }.map { it as Annotation }
        val name = ctx.textAt(annotations.size)
        val datatype = list[annotations.size] as ISerializable
        val options = list.firstOrNull { it is FieldOptions } ?: FieldOptions(FieldOptions.FieldOptionsOptions(false))
        val defaultVal = list.filter { it is Value }.map { (it as Value).options.value }.firstOrNull()
        return listOf(FieldType(FieldType.FieldTypeOptions(name, annotations, datatype, options as FieldOptions, defaultVal)))
    }

    override fun visitDataType(ctx: DynabuffersParser.DataTypeContext): List<IType> {
        return when (ctx.text) {
            "string" -> listOf(StringType(StringType.StringTypeOptions(charset)))
            "float" -> listOf(FloatType())
            "int" -> listOf(IntType())
            "boolean" -> listOf(BooleanType())
            "long" -> listOf(LongType())
            "byte" -> listOf(ByteType())
            "short" -> listOf(ShortType())
            "map" -> listOf(MapType(MapType.MapTypeOptions(charset)))
            else -> listOf(RefType(RefType.RefTypeOptions(ctx.text)))
        }
    }

    override fun visitArrayType(ctx: DynabuffersParser.ArrayTypeContext): List<IType> {
        val datatype = super.visitArrayType(ctx)[0] as ISerializable
        return listOf(ArrayType(ArrayType.ArrayTypeOptions(datatype)))
    }

    override fun visitOptionType(ctx: DynabuffersParser.OptionTypeContext?): List<IType> {
        val datatype = super.visitOptionType(ctx)[0] as ISerializable
        return listOf(OptionType(OptionType.OptionTypeOptions(datatype)))
    }

    override fun visitClassType(ctx: DynabuffersParser.ClassTypeContext): List<IType> {
        val list = super.visitClassType(ctx)
        val classOptions = list.firstOrNull { it is ClassOptions }
                ?: ClassOptions(ClassOptions.ClassOptionsOptions(false, false, false))

        val fields = list.filter { it is FieldType }.map { it as FieldType }
        val options = ClassType.ClassTypeOptions(ctx.textAt(1), fields, classOptions as ClassOptions)

        return listOf(ClassType(options))
    }

    override fun visitAnnotation(ctx: DynabuffersParser.AnnotationContext): List<IType> {
        val name = ctx.textAt(1)
        val values = super.visitAnnotation(ctx).map { it as Value }
        return listOf(Annotation(Annotation.AnnotationTypeOptions(name, values)))
    }

    override fun visitValue(ctx: DynabuffersParser.ValueContext): List<IType> {
        return listOf(Value(Value.ValueTypeOptions(when (ctx.text) {
            "[]" -> ArrayList<Any>()
            "[:]" -> HashMap<String, Any>()
            else -> ctx.text
        })))
    }

    override fun visitClassOptions(ctx: DynabuffersParser.ClassOptionsContext): List<IType> {
        return listOf(ClassOptions(ClassOptions.ClassOptionsOptions(
                ctx.text.contains("primary"),
                ctx.text.contains("deprecated"),
                ctx.text.contains("implicit"))))
    }

    override fun visitUnionOptions(ctx: DynabuffersParser.UnionOptionsContext): List<IType> {
        return listOf(UnionOptions(UnionOptions.UnionOptionsOptions(
                ctx.text.contains("primary"),
                ctx.text.contains("deprecated"),
                ctx.text.contains("implicit"))))
    }

    override fun visitFieldOptions(ctx: DynabuffersParser.FieldOptionsContext): List<IType> {
        return listOf(FieldOptions(FieldOptions.FieldOptionsOptions(ctx.text.contains("deprecated"))))
    }

    override fun visitNamespaceType(ctx: DynabuffersParser.NamespaceTypeContext): List<IType> {
        val name = ctx.getChild(1).text
        val list = super.visitNamespaceType(ctx).map { it as ISerializable }
        return listOf(NamespaceType(NamespaceType.NamespaceTypeOptions(name, list)))
    }

    override fun aggregateResult(aggregate: List<IType>?, nextResult: List<IType>?): List<IType> {
        val list = ArrayList<IType>()
        if (aggregate != null) list.addAll(aggregate)
        if (nextResult != null) list.addAll(nextResult)
        return list
    }

    private fun ParserRuleContext.textAt(index: Int) = this.getChild(index).text
    private fun ParserRuleContext.textAt(left: Int, right: Int) = (left..this.childCount - right).map { this.textAt(it) }

}
